File: hooks/scripts/cmd/tool_output_truncation.go
// PostToolUse hook: truncates large tool outputs to save context window space.
package cmd

import (
	"os"

	"github.com/spf13/cobra"
	"solon-hooks/internal/config"
	"solon-hooks/internal/hookio"
	"solon-hooks/internal/truncation"
)

// truncationWhitelist contains tools whose output must never be truncated.
// Claude needs full output from these to verify file changes correctly.
var truncationWhitelist = map[string]bool{
	"Edit":         true,
	"Write":        true,
	"MultiEdit":    true,
	"NotebookEdit": true,
}

var toolOutputTruncationCmd = &cobra.Command{
	Use:   "tool-output-truncation",
	Short: "Truncate large tool outputs to save context",
	RunE:  runToolOutputTruncation,
}

func runToolOutputTruncation(cmd *cobra.Command, args []string) error {
	if !config.IsHookEnabled("tool-output-truncation") {
		os.Exit(0)
	}

	var input hookio.PostToolUseInput
	if err := hookio.ReadInput(&input); err != nil {
		hookio.WriteOutput(map[string]interface{}{"continue": true})
		return nil
	}

	if truncationWhitelist[input.ToolName] {
		hookio.WriteOutput(map[string]interface{}{"continue": true})
		return nil
	}

	budget := truncation.BudgetForTool(input.ToolName)
	result, changed := truncation.TruncateOutput(
		input.ToolOutput,
		budget.MaxLines,
		budget.HeadLines,
		budget.TailLines,
	)

	if !changed {
		hookio.WriteOutput(map[string]interface{}{"continue": true})
		return nil
	}

	hookio.WriteOutput(map[string]interface{}{
		"continue": true,
		"hookSpecificOutput": map[string]interface{}{
			"tool_output": result,
		},
	})
	return nil
}

File: crates/solon-core/src/hooks/tool_output_truncation.rs
use crate::hooks::{budget_for_tool, get_str, is_hook_enabled, truncate_output};
/// PostToolUse hook: truncate large tool outputs to save context window space.
use anyhow::Result;

/// Tools whose output must never be truncated — Claude needs full content to verify file changes.
const TRUNCATION_WHITELIST: &[&str] = &["Edit", "Write", "MultiEdit", "NotebookEdit"];

pub fn run() -> Result<()> {
    if !is_hook_enabled("tool-output-truncation") {
        std::process::exit(0);
    }

    let input = match crate::hooks::read_hook_input() {
        Ok(v) => v,
        Err(_) => {
            crate::hooks::write_output(&serde_json::json!({"continue": true}));
            return Ok(());
        }
    };

    let tool_name = get_str(&input, "tool_name").to_string();
    if TRUNCATION_WHITELIST.contains(&tool_name.as_str()) {
        crate::hooks::write_output(&serde_json::json!({"continue": true}));
        return Ok(());
    }

    let tool_output = input
        .get("tool_output")
        .and_then(|v| v.as_str())
        .unwrap_or("")
        .to_string();

    let budget = budget_for_tool(&tool_name);
    let (result, changed) = truncate_output(
        &tool_output,
        budget.max_lines,
        budget.head_lines,
        budget.tail_lines,
    );

    if !changed {
        crate::hooks::write_output(&serde_json::json!({"continue": true}));
        return Ok(());
    }

    crate::hooks::write_output(&serde_json::json!({
        "continue": true,
        "hookSpecificOutput": {
            "tool_output": result,
        }
    }));
    Ok(())
}
